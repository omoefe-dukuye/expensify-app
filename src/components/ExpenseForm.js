import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends Component {
  existingState = () => {
    const { expense } = this.props;
    return expense && {
      ...expense,
      amount: (expense.amount / 100).toString(),
      createdAt: moment(expense.createdAt)
    };
  };

  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    ...this.existingState(),
    calendarFocused: false,
    error: '',
  };

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (amount.match(/^\d*(\d+\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => createdAt && this.setState(() => ({ createdAt }));

  onFocusChange = ({ focused: calendarFocused }) => {
    this.setState(() => ({ calendarFocused }));
  };

  onSubmit = e => {
    const {
      state: { description, amount, createdAt, note },
      props,
    } = this;
    e.preventDefault();
    if (description && amount) {
      this.setState(() => ({ error: '' }));
      props.onSubmit({
        description,
        amount: Number(amount) * 100,
        createdAt: createdAt.valueOf(),
        note,
      });
    } else {
      this.setState(() => ({ error: 'Please enter description and amount' }));
    }
  };

  render() {
    const {
      state: { description, note, amount, createdAt, calendarFocused, error },
      onAmountChange,
      onDescriptionChange,
      onNoteChange,
      onDateChange,
      onFocusChange,
      onSubmit,
    } = this;
    return (
      <div>
        {error && <p>{error}</p>}
        <form onSubmit={onSubmit}>
          <input type="text" placeholder="Description" onChange={onDescriptionChange} value={description} />
          <input type="text" placeholder="Amount" value={amount} onChange={onAmountChange} />
          <SingleDatePicker
            date={createdAt}
            onDateChange={onDateChange}
            focused={calendarFocused}
            onFocusChange={onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add note for your expense (Optional)"
            onChange={onNoteChange}
            value={note}
          />
          <button type="submit">{`${this.existingState() ? 'Edit' : 'Add'} Expense`}</button>
        </form>
      </div>
    );
  }
}
