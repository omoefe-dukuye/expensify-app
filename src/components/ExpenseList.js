import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

const expenseArray = expenses => (
  expenses.map(expense => <ExpenseListItem key={expense.id} {...expense}/>)
);

export const ExpenseList = ({ expenses }) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>

    <div className="list-body">
      {
        expenses.length > 0 ? (
          expenseArray(expenses)
        ) : (
          <div className="list-item list-item__message">
            <span>No expenses</span>
          </div>
        )
      }
    </div>
  </div>
);

const mapStateToProps = ({ expenses, filters }) => ({
  expenses: selectExpenses(expenses, filters)
});

export default connect(mapStateToProps)(ExpenseList);
