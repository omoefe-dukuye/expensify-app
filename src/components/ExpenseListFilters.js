import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import {
  setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate
} from '../actions/filters';


export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onTextFilterChange = ({ target: { value: text } }) => {
    this.props.setTextFilter(text)
  };

  onSortByChange = ({ target: { value: sortBy } }) => {
    sortBy === 'date'
      ? this.props.sortByDate()
      : this.props.sortByAmount()
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              onChange={this.onTextFilterChange}
              placeholder="Search expenses"
              type="text" value={this.props.filters.text}
            />
          </div>

          <div className="input-group__item">
            <select className="select" onChange={this.onSortByChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>

          <div className="input-group__item">
            <DateRangePicker
              startDateId="startDateId"
              startDate={this.props.filters.startDate}
              endDateId="endDateId"
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              showClearDates={true}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ filters }) => ({ filters });

const mapDispatchToProps = (dispatch) => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);