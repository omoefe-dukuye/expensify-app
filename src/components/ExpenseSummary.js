import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import number from '../util/numeral-locale-naira';
import selector from '../selectors/expenses';
import amountCalculator from '../selectors/expense-total';

export const ExpenseSummary = ({ expenses, filters }) => {
  const filterdExpenses = selector(expenses, filters);
  const expenseCount = filterdExpenses.length;
  const totalAmount = amountCalculator(filterdExpenses);
  const totalAmountInNaira = number(totalAmount).divide(100).format('$0,0.00');
  
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount}</span>&nbsp;
          expense{expenseCount !== 1 && 's'}&nbsp;
          totalling <span>{totalAmountInNaira}</span>
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = ({ expenses, filters }) => ({ expenses, filters });

export default connect(mapStateToProps)(ExpenseSummary);
