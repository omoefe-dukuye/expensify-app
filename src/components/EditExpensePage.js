import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  editExpense = (expense) => {
    this.props.editExpense(expense);
    this.props.history.push('/');
  };

  removeExpense = () => {
    this.props.removeExpense();
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.editExpense} />
          <button className="button button--grey" onClick={this.removeExpense}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ expenses }, props) => ({
  expense: expenses.find(({ id }) => id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, { match: { params: { id } } }) => ({
  editExpense: expense => dispatch(startEditExpense(id, expense)),
  removeExpense: () => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
