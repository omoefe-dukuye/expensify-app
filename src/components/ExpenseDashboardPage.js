import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseSummary from './ExpenseSummary';

export default () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);