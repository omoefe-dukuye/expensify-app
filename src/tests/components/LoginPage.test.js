import React from 'react';
import { LoginPage } from '../../components/LoginPage';
import { shallow } from 'enzyme';

test('Should render login page', () => {
  const wrapper = shallow(<LoginPage startLogin={() => {}}/>);
  expect(wrapper).toMatchSnapshot();
});

test('Should call startLogin prop on click', () => {
  const startLogin = jest.fn();
  const wrapper = shallow(<LoginPage startLogin={startLogin}/>);
  wrapper.find('button').simulate('click');
  expect(startLogin).toHaveBeenCalled();
});
