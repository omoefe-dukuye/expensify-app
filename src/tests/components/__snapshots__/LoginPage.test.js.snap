// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render login page 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Log in
  </button>
</div>
`;
